From 5c3cd662e220682d06f18a9873eb97649aa0728c Mon Sep 17 00:00:00 2001
From: Yufen Yu <yuyufen@huawei.com>
Date: Tue, 19 Mar 2019 07:23:35 +0200
Subject: [PATCH 2/2] block: Fix use-after-free of gendisk

Patch 1: fix use after free in get_gendisk, when open device partition.
Patch 2: remove unneccessary if statement

Yufen Yu (2):
  block: remove devt from ext_devt_idr when delete partition
  block: remove unnecessary statement in blk_free_devt

Signed-off-by: Yufen Yu <yuyufen@huawei.com>
---
 block/genhd.c             | 3 ---
 block/partition-generic.c | 5 +++++
 2 files changed, 5 insertions(+), 3 deletions(-)

diff --git a/block/genhd.c b/block/genhd.c
index dc111f781e45..3d70ac1d7003 100644
--- a/block/genhd.c
+++ b/block/genhd.c
@@ -523,9 +523,6 @@ int blk_alloc_devt(struct hd_struct *part, dev_t *devt)
  */
 void blk_free_devt(dev_t devt)
 {
-	if (devt == MKDEV(0, 0))
-		return;
-
 	if (MAJOR(devt) == BLOCK_EXT_MAJOR) {
 		spin_lock_bh(&ext_devt_lock);
 		idr_remove(&ext_devt_idr, blk_mangle_minor(MINOR(devt)));
diff --git a/block/partition-generic.c b/block/partition-generic.c
index 2bade849cc5c..3de5139f2e1d 100644
--- a/block/partition-generic.c
+++ b/block/partition-generic.c
@@ -285,6 +285,11 @@ void delete_partition(struct gendisk *disk, int partno)
 	kobject_put(part->holder_dir);
 	device_del(part_to_dev(part));
 
+	/*
+	 * We should ensuere to delete part from idr before kfree(disk),
+	 * avoiding use-after-free of disk.
+	 */
+	blk_free_devt(part_devt(part));
 	hd_struct_kill(part);
 }
 
-- 
2.21.0.155.ge902e9bcae

